src/gui/editors/PianoRoll.cpp

```diff
@@ -1458,9 +1458,47 @@ void PianoRoll::keyPressEvent(QKeyEvent* ke)
 
 		case Qt::Key_Backspace:
 		case Qt::Key_Delete:
-			deleteSelectedNotes();
-			ke->accept();
-			break;
+			{
+				// Get start pos and end pos of selection; used for shift + delete
+				auto selectedNotes = getSelectedNotes();
+				if (selectedNotes.empty()) { break; }
+				TimePos delStartPos = selectedNotes.first()->pos();
+				TimePos delEndPos = selectedNotes.last()->endPos();
+				TimePos delEndLen = selectedNotes.last()->length();
```

Comment:
A prior note may still end after the last note depending on its length. You may find `boundsForNotes` helpful.

It will still be tricky with drum notes, but I imagine something like this:

```
area to delete     |xxxxxx|            |xxxxxxxxx|
normal note         ######    ####      ####      ###
drum note              #                    #  
```

In that case you just need to check if the drum note starts on the end position... I also see that `boundsForNotes` doesn't handle drum notes well. That needs to be fixed somehow.